RDP session closes when the session lock screen appears. Seems related to the recent clipboard updates.

Hello XRDP development team,

I currently use XRDP to connect to existing VNC sessions. So I encountered the copy/paste problem for messages that close RDP session.
While browsing the XRDP GitHub, I saw that this issue has been fixed and reviewed recently but there is no official release yet.
So, I downloaded the available source code and I made the installation from the source following the instructions described.

At first sight, the copy and paste function seems to work correctly from the Windows built-in RDP client. 

However, this update seems to release a new problem.
Every time my session locks after a few minutes of inactivity or I lock my session manually. As soon as I move the mouse back into the window to access it again, it closes instantly.

In response to this event, I found this in the logfile:

[*DATE*-*TIME*] [DEBUG] VNC got clip data
[*DATE*-*TIME*] [ERROR] Memory exhausted allocating 0 bytes for RFB clip data
[*DATE*-*TIME*] [ERROR] lib_data_in: lib_mod_process_message failed
[*DATE*-*TIME*] [DEBUG] Closed socket 12 (AF_INET **RDP_SERVER_IP**:**PORT**)
[*DATE*-*TIME*] [DEBUG] xrdp_mm_module_cleanup
[*DATE*-*TIME*] [DEBUG] Closed socket 18 (AF_INET **VNC_SERVER_IP**:**PORT**)

Would you have a solution to solve this problem?

I can also share any additional information needed to debug this issue.

Thanks !

--- a/vnc/vnc_clip.c
+++ b/vnc/vnc_clip.c
@@ -1003,21 +1003,34 @@ vnc_clip_process_rfb_data(struct vnc *v)
             vc->rfb_clip_s = 0;
 
             make_stream(vc->rfb_clip_s);
-            if (size >= 0)
+            if (size < 0)
             {
-                init_stream(vc->rfb_clip_s, size);
-            }
-            if (vc->rfb_clip_s->data == NULL)
-            {
-                LOG(LOG_LEVEL_ERROR,
-                    "Memory exhausted allocating %d bytes for RFB clip data",
-                    size);
+                /* This shouldn't happen - see Extended Clipboard
+                 * Pseudo-Encoding */
+                LOG(LOG_LEVEL_ERROR, "Unexpected size %d for RFB data", size);
                 rv = 1;
+            }
+            else if (size == 0)
+            {
+                LOG(LOG_LEVEL_DEBUG, "RFB clip data cleared by VNC server");
             }
             else
             {
-                LOG(LOG_LEVEL_DEBUG, "Reading %d clip bytes from RFB", size);
-                rv = trans_force_read_s(v->trans, vc->rfb_clip_s, size);
+                init_stream(vc->rfb_clip_s, size);
+                if (vc->rfb_clip_s->data == NULL)
+                {
+                    LOG(LOG_LEVEL_ERROR,
+                        "Memory exhausted allocating %d bytes"
+                        " for RFB clip data",
+                        size);
+                    rv = 1;
+                }
+                else
+                {
+                    LOG(LOG_LEVEL_DEBUG, "Reading %d clip bytes from RFB",
+                        size);
+                    rv = trans_force_read_s(v->trans, vc->rfb_clip_s, size);
+                }
             }
 
             /* Consider telling the RDP client about the update only if we've

